src/parser.c

```diff
@@ -0,0 +1,164 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   parser.c                                           :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: sqiu <sqiu@student.42vienna.com>           +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/06/23 13:13:28 by sqiu              #+#    #+#             */
+/*   Updated: 2023/07/29 13:33:56 by sqiu             ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file parser.c
+ * @brief Contains functions to parse the token list, categorise the tokens and
+ * create a cmd list.
+ */
+
+#include "mod_parser.h"
+
+/**
+ * @brief 			Driver function to parse token list.
+ * 
+ * Creates cmd structures and fills them with the token list.
+ * @param lst 		Pointer to list of tokens.
+ * @param cmd_head 	Pointer pointer to first node of cmd list.
+ * @return t_err 	ERR_MALLOC, SUCCESS
+ */
+t_err	ft_parser(t_tkn_list *lst, t_cmd **cmd_head)
+{
+	t_err	err;
+	t_cmd	*new;
+	bool	cmd_complete;
+
+	cmd_complete = 0;
+	new = NULL;
+	err = ft_create_cmd(&new, lst);
+	if (err != SUCCESS)
+		return (err);
+	while (lst)
+	{
+		err = ft_categorise(&lst, new, &cmd_complete);
+		if (err != SUCCESS)
+			return (err);
+		if (cmd_complete)
+			new = ft_lock_and_load_cmd(lst, new, cmd_head, &cmd_complete);
+		if (!new)
+			return (ERR_MALLOC);
+		lst = lst->next;
+	}
+	ft_add_cmd(new, cmd_head);
+	return (SUCCESS);
+}
+
+/**
+ * @brief Adds current cmd to cmd_list and creates a new cmd struct.
+ * 
+ * @param lst			List of tokens at end position of previous cmd.
+ * @param curr 			Current cmd struct.
+ * @param cmd_head 		List of cmd structs.
+ * @param cmd_complete 	Bool to indicate if cmd ist complete.
+ * @return t_cmd* 
+ */
+t_cmd	*ft_lock_and_load_cmd(t_tkn_list *lst, t_cmd *curr, t_cmd **cmd_head,
+		bool *cmd_complete)
+{
+	t_cmd	*new;
+	t_err	err;
+
+	new = NULL;
+	ft_add_cmd(curr, cmd_head);
+	lst = lst->next;
+	err = ft_create_cmd(&new, lst);
+	if (err != SUCCESS)
+		return (NULL);
+	*cmd_complete = 0;
+	return (new);
+}
+
+/**
+ * @brief Initiates the cmd struct.
+ * 
+ * @param new 		Newly created cmd struct.
+ * @param lst		List of tokens at current position.
+ * @return t_err 	ERR_MALLOC, SUCCESS
+ */
+t_err	ft_create_cmd(t_cmd **new, t_tkn_list *lst)
+{
+	t_cmd	*tmp;
+	t_err	err;
+	int		count_arg;
+	int		count_delim;
+
+	count_arg = 0;
+	count_delim = 0;
+	ft_count_str(lst, &count_arg, &count_delim);
+	err = ft_create_str_arr(&tmp, count_arg, count_delim);
```

Comment:
I would leave the malloc() of the t_cmd itself out of the ft_create_str_arr(), because it hides the mallocing a bit.
But it seems like a line limit problem. Alternative would be to refactor the initilisation step itself (set stuff to null) into a ft_init_cmd() and then just call the function.

Reply:
It is not (only) a line issue rather than yet again a pointer issue. As I am assigning adresses in ft_create_str_arr() which I'd like to carry back to ft_create_cmd() I need to pass the pointer pointer. That in itself requires me to use a proxy t_cmd *tmp in ft_create_str_arr(). If I don't malloc that proxy, I will get a core dump while compiling. If I am already mallocing, why not just use it for the actual t_cmd that I want to continue using.

Reply:
Why would you need to malloc a proxy? Since you already have a struct, which let's you pass the pointers directly you don't even need a pointer pointer.

- You can pass the newly allocated t_cmd as t_cmd *tmp
- Then you just save the addresses of the newly malloced arrays in tmp->delim or tmp->args (you can directly set the address. This is because tmp->arg is not a function copy. The "pointer pointer" in this case is *tmp)

Reply:
These pointer pointers and I will never be friends. Changed, tested, works.